[torchdim] Fix Python 3.11 bytecode decoding in dims

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91290

Adds a PyInstDecoder object that handles the differences in bytecode
added in 3.11. Basically some instructions have inline caches which
change the size of the instruction, so calculating the next instruction
is slightly different.

fixes #91246

zdevito committed:
[torchdim] Fix Python 3.11 bytecode decoding in dims

Adds a PyInstDecoder object that handles the differences in bytecode
added in 3.11. Basically some instructions have inline caches which
change the size of the instruction, so calculating the next instruction
is slightly different.

[ghstack-poisoned]

albanD commented:
Dont hesitate to use https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/python_compat.h and https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/pythoncapi_compat.h to reduce the need to ifdef. The functions there also handle the change in behavior of some functions from one version to the next (like lasti).
These are pure C headers so will be safe to include anywhere.

zdevito committed:
Update on "[torchdim] Fix Python 3.11 bytecode decoding in dims"


Adds a PyInstDecoder object that handles the differences in bytecode
added in 3.11. Basically some instructions have inline caches which
change the size of the instruction, so calculating the next instruction
is slightly different.

fixes #91246

[ghstack-poisoned]

zdevito commented:
Updated to remove the ifdefs possible. Looks like it passes on old pythons on the CI, and I verified locally it works on 3.11 (CPU).

albanD commented on functorch/csrc/dim/dim.cpp:
```diff
@@ -21,6 +21,11 @@
 #include "arena.h"
 #include "python_variable_simple.h"
 
+#if PY_VERSION_HEX >= 0x030b0000
```
nit: You have a `IS_PYTHON_3_11_PLUS`

albanD commented on functorch/csrc/dim/dim.cpp:
```diff
@@ -1491,6 +1469,50 @@ py::object create_dimlist(py::object name, py::handle size) {
+                names = py::object::steal(PyCode_GetCellvars(code_object_));
+                break;
+            default:
+                return py::object();
```
Do you actually want to return an empty object?
```suggestion
                Py_RETURN_NONE

albanD reviewed (approved):
Sounds good!

zdevito committed:
Update on "[torchdim] Fix Python 3.11 bytecode decoding in dims"


Adds a PyInstDecoder object that handles the differences in bytecode
added in 3.11. Basically some instructions have inline caches which
change the size of the instruction, so calculating the next instruction
is slightly different.

fixes #91246

[ghstack-poisoned]

zdevito commented:
@pytorchbot rebase

zdevito committed:
Update on "[torchdim] Fix Python 3.11 bytecode decoding in dims"


Adds a PyInstDecoder object that handles the differences in bytecode
added in 3.11. Basically some instructions have inline caches which
change the size of the instruction, so calculating the next instruction
is slightly different.

fixes #91246

[ghstack-poisoned]

albanD commented:
@pytorchbot merge -r

zdevito committed:
Update on "[torchdim] Fix Python 3.11 bytecode decoding in dims"


Adds a PyInstDecoder object that handles the differences in bytecode
added in 3.11. Basically some instructions have inline caches which
change the size of the instruction, so calculating the next instruction
is slightly different.

fixes #91246

[ghstack-poisoned]

albanD commented:
@pytorchbot merge -f "Flaky rocm"
